README
NAME
    Stump - Larry Wall's Slideshow Software

SYNOPSIS
        > stump init
        > edit stump.input
        > stump make
        > stump speech

DESCRIPTION
    Stump is Larry Wall's slideshow presentation hacks, packaged up for
    CPAN.

    Stump takes a simple input format and some pictures and whatnot, and
    compiles them into a OpenDocument (.odp) file that you view with your
    favorite slideshow software.

AUTHOR
    Larry Wall via Ingy döt Net <ingy@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (c) 2011. Larry Wall via Ingy döt Net.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html



Changes
---
version: 0.13
date:    Tue Aug 19 16:32:39 PDT 2014
changes:
- Change tabs to spaces
---
version: 0.12
date:    Sat Aug 16 12:22:42 PDT 2014
changes:
- Meta 0.0.2
---
version: 0.11
date:    Fri Aug 15 21:29:12 PDT 2014
changes:
- Remove t/use_ok.t
---
version: 0.10
date:    Fri Aug 15 20:39:24 PDT 2014
changes:
- Add t/000-compile-modules.t
---
version: 0.09
date:    Fri Aug  8 12:57:57 PDT 2014
changes:
- Switch hashbang to '#!/usr/bin/env perl'
---
version: 0.08
date:    Wed Jul 30 16:19:09 PDT 2014
changes:
- Switch IRC to '#pkg'
---
version: 0.07
date:    Thu Jul 24 23:39:35 PDT 2014
changes:
- Switch to Zilla-Dist
---
version: 0.06
date:    Sat Aug 27 14:36:31 CEST 2011
changes:
- Require MouseX::App::Cmd
---
version: 0.05
date:    Fri Aug 26 01:52:50 CEST 2011
changes:
- Require Mouse
---
version: 0.04
date:    Mon Aug 22 11:10:02 CEST 2011
changes:
- Require App::Cmd
---
version: 0.03
date:    Wed Aug 17 11:57:51 EEST 2011
changes:
- Remove conf as option
---
version: 0.02
date:    Mon Aug 15 15:30:53 EEST 2011
changes:
- Make it actually work...
---
version: 0.01
date:    Mon Aug 15 13:04:38 EEST 2011
changes:
- First Release


lib/Stump.pm
use strict; use warnings;
package Stump;
our $VERSION = '0.12';

use Mouse;
use MouseX::App::Cmd;
use App::Cmd;
use File::Share;
use IO::All;
use YAML::XS;

#-----------------------------------------------------------------------------#
package Stump::Command;
use App::Cmd::Setup -command;
use Mouse;
extends qw[MouseX::App::Cmd::Command];

sub validate_args {
}

# Semi-brutal hack to suppress extra options I don't care about.
around usage => sub {
    my $orig = shift;
    my $self = shift;
    my $opts = $self->{usage}->{options};
    @$opts = grep { $_->{name} ne 'help' } @$opts;
    return $self->$orig(@_);
};

#-----------------------------------------------------------------------------#
package Stump;
use App::Cmd::Setup -app;
use Mouse;
extends 'MouseX::App::Cmd';

use Module::Pluggable
  require     => 1,
  search_path => [ 'Stump' ];
Stump->plugins;

# App::Cmd help helpers
use constant usage => 'Stump';
use constant text => "stump command [<options>] [<arguments>]\n";

#-----------------------------------------------------------------------------#
package Stump::Command::init;
Stump->import( -command );
use Mouse;
extends qw[Stump::Command];

use constant abstract => 'Initialize a new Stump presentation';
use constant usage_desc => 'stump init [--force]';

has force => (
    is => 'ro',
    isa => 'Bool',
    documentation => 'Force an init operation',
);

sub execute {
    my ($self, $opt, $args) = @_;

    if ($self->empty_directory or $self->force) {
        my $share = $self->share;
        $self->copy_file("$share/stump.input", "./stump.input");
        $self->copy_file("$share/conf.yaml", "./conf.yaml");
        $self->copy_files("$share/image", "./image");
        $self->init_ok_msg;
    }
    else {
        $self->error__wont_init;
    }
}

#-----------------------------------------------------------------------------#
package Stump::Command::make;
Stump->import( -command );
use Mouse;
extends qw[Stump::Command];

use IO::All;

use constant abstract => 'Make a Stump ODP Presentation';
use constant usage_desc => 'stump make';

sub execute {
    my ($self, $opt, $args) = @_;
    $self->make;
}

#-----------------------------------------------------------------------------#
package Stump::Command::speech;
Stump->import( -command );
use Mouse;
extends qw[Stump::Command];

use constant abstract => 'Start your Stump speech';
use constant usage_desc => 'stump speech';

sub execute {
    my ($self, $opt, $args) = @_;

    my $start_command = $self->conf->{start_command}
        or die "No 'start_command' in conf.yaml";

    $self->make
        if $self->conf->{auto_make} and (
            not -e 'stump.odp' or
            -M 'stump.input' < -M 'stump.odp'
        );

    exec $start_command;
}

#-----------------------------------------------------------------------------#
package Stump::Command::clean;
Stump->import( -command );
use Mouse;
extends qw[Stump::Command];

use constant abstract => 'Cleanup generated files';
use constant usage_desc => 'stump clean';

sub execute {
    my ($self, $opt, $args) = @_;
    system('rm -fr stump stump.odp');
}

#-----------------------------------------------------------------------------#
# Helper methods
#-----------------------------------------------------------------------------#
package Stump::Command;
use File::Share;
use IO::All;
use Cwd qw[cwd abs_path];
use YAML::XS;

has _conf => (
    reader => 'conf',
    is => 'ro',
    lazy => 1,
    builder => sub {
        my $self = shift;
        my $file = 'conf.yaml';
        die "There is no conf.yaml file.\n"
            unless -e $file;
        YAML::XS::LoadFile($file);
    },
);

sub make {
    require Stump::Heavy;
    my ($self) = @_;
    my $share = $self->share;
    $self->copy_file("$share/stump.odp", "./stump.odp");
    Stump::Heavy::para2odp();
    io('stump')->rmtree;
}

sub share {
    File::Share::dist_dir('Stump');
}

sub empty_directory {
    io('.')->empty;
}

sub copy_file {
    my ($self, $source, $target) = @_;
    my $file = io($source);
    io("$target")->assert->print($file->all);
}

sub copy_files {
    my ($self, $source, $target) = @_;
    for my $file (io($source)->All_Files) {
        my $short = $file->name;
        $short =~ s!^\Q$source\E/?!! or die $short;
        next if $short =~ /^\./;
        io("$target/$short")->assert->print($file->all);
    }
}

sub init_ok_msg {
    print <<'...';

Stump slideshow created.

Now edit 'stump.input' and 'conf.yaml' and run:

    stump make
    stump speech

...
}

sub error {
    my ($self, $msg) = splice(@_, 0, 2);
    chomp $msg;
    $msg .= $/;
    die sprintf($msg, @_);
}

sub error__wont_init {
    my ($self) = @_;
    $self->error(
        "Won't 'init' in a non empty directory, unless you use --force"
    );
}

1;


lib/Stump/Heavy.pm
use strict; use warnings;
package Stump::Heavy;

# What follows is an ongoing refactoring of Larry's original code as given to
# Ingy @ YAPC::Riga 2011:

use utf8;

my %COLOR = qw(
    S Red
    E Purple
    B Blue
    G Green
    C Cyan
    M Magenta
    Y Yellow
    O Orange
    R Red
    L Lavender
    P Purple
    K Black
    g Gray
    W White
);

binmode(STDIN, ':utf8');
binmode(STDOUT, ':utf8');
binmode(STDERR, ':utf8');


sub para2odp {

my $input = 'stump.input';
my $build = 'stump';
my $template = 'stump.odp';

die "Invalid template" unless -e $template;
system "mkdir $build; cd $build; unzip ../$template"
    unless -d $build;

open CONTENT, ">:utf8", "$build/content.xml"
    or die "Can't create $build/content.xml: $!";
system "rm -rf $build/Pictures/*";

my $text;
my @para;
{
    open IN, '<:utf8', $input or die "Can't open $input: $!";
    local $/ = "";
    chomp(@para = <IN>);
    close IN;
}

print CONTENT <<'END';
<?xml version="1.0" encoding="UTF-8"?>

<office:document-content xmlns:office='urn:oasis:names:tc:opendocument:xmlns:office:1.0' xmlns:rpt='http://openoffice.org/2005/report' grddl:transformation='http://docs.oasis-open.org/office/1.2/xslt/odf2rdf.xsl' xmlns:math='http://www.w3.org/1998/Math/MathML' xmlns:field='urn:openoffice:names:experimental:ooo-ms-interop:xmlns:field:1.0' xmlns:ooo='http://openoffice.org/2004/office' xmlns:form='urn:oasis:names:tc:opendocument:xmlns:form:1.0' xmlns:xsd='http://www.w3.org/2001/XMLSchema' xmlns:of='urn:oasis:names:tc:opendocument:xmlns:of:1.2' xmlns:dom='http://www.w3.org/2001/xml-events' xmlns:oooc='http://openoffice.org/2004/calc' xmlns:style='urn:oasis:names:tc:opendocument:xmlns:style:1.0' xmlns:script='urn:oasis:names:tc:opendocument:xmlns:script:1.0' xmlns:presentation='urn:oasis:names:tc:opendocument:xmlns:presentation:1.0' xmlns:tableooo='http://openoffice.org/2009/table' xmlns:css3t='http://www.w3.org/TR/css3-text/' xmlns:grddl='http://www.w3.org/2003/g/data-view#' xmlns:smil='urn:oasis:names:tc:opendocument:xmlns:smil-compatible:1.0' xmlns:dr3d='urn:oasis:names:tc:opendocument:xmlns:dr3d:1.0' xmlns:text='urn:oasis:names:tc:opendocument:xmlns:text:1.0' office:version='1.2' xmlns:number='urn:oasis:names:tc:opendocument:xmlns:datastyle:1.0' xmlns:ooow='http://openoffice.org/2004/writer' xmlns:xforms='http://www.w3.org/2002/xforms' xmlns:xlink='http://www.w3.org/1999/xlink' xmlns:formx='urn:openoffice:names:experimental:ooxml-odf-interop:xmlns:form:1.0' xmlns:anim='urn:oasis:names:tc:opendocument:xmlns:animation:1.0' xmlns:table='urn:oasis:names:tc:opendocument:xmlns:table:1.0' xmlns:xhtml='http://www.w3.org/1999/xhtml' xmlns:officeooo='http://openoffice.org/2009/office' xmlns:draw='urn:oasis:names:tc:opendocument:xmlns:drawing:1.0' xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance' xmlns:meta='urn:oasis:names:tc:opendocument:xmlns:meta:1.0' xmlns:fo='urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0' xmlns:svg='urn:oasis:names:tc:opendocument:xmlns:svg-compatible:1.0' xmlns:chart='urn:oasis:names:tc:opendocument:xmlns:chart:1.0' xmlns:dc='http://purl.org/dc/elements/1.1/'>
  <office:scripts />
  <office:automatic-styles>
    <style:style style:name='dp1' style:family='drawing-page'>
      <style:drawing-page-properties presentation:background-visible='true' presentation:display-date-time='true' draw:fill='solid' draw:fill-image-width='0cm' draw:fill-color='#ffffff' presentation:display-page-number='false' presentation:display-footer='true' draw:fill-image-height='0cm' presentation:background-objects-visible='true' />
    </style:style>
    <style:style style:name='dp2' style:family='drawing-page'>
      <style:drawing-page-properties presentation:display-date-time='true' presentation:display-page-number='false' presentation:display-footer='true' presentation:display-header='true' />
    </style:style>
    <style:style style:name='dp3' style:family='drawing-page'>
      <style:drawing-page-properties presentation:background-visible='true' presentation:display-date-time='true' presentation:display-page-number='false' presentation:display-footer='true' presentation:background-objects-visible='true' />
    </style:style>
    <style:style style:parent-style-name='standard' style:name='middling' style:family='graphic'>
      <style:graphic-properties svg:stroke-color='#000000' draw:stroke='none' fo:max-height='0cm' draw:fill='none' draw:fill-color='#ffffff' draw:auto-grow-height='true' draw:textarea-vertical-align='middle' fo:min-height='0cm' draw:auto-grow-width='false' />
    </style:style>
    <style:style style:name='gr2' style:family='graphic'>
      <style:graphic-properties style:protect='size' />
    </style:style>
    <style:style style:parent-style-name='standard' style:name='takahashi' style:family='graphic'>
      <style:graphic-properties svg:stroke-color='#000000' draw:stroke='none' fo:min-width='3.296cm' draw:textarea-horizontal-align='center' draw:textarea-vertical-align='middle' draw:fill='none' draw:fill-color='#ffffff' draw:auto-grow-height='true' fo:min-height='1.578cm' draw:auto-grow-width='true' />
    </style:style>
    <style:style style:parent-style-name='standard' style:name='gr4' style:family='graphic'>
      <style:graphic-properties draw:red='0%' draw:stroke='none' draw:contrast='0%' draw:textarea-horizontal-align='center' draw:blue='0%' draw:fill='none' fo:clip='rect(0cm, 0cm, 0cm, 0cm)' draw:textarea-vertical-align='middle' draw:gamma='100%' draw:luminance='0%' draw:green='0%' draw:color-mode='standard' draw:image-opacity='100%' style:mirror='none' />
    </style:style>
    <style:style style:parent-style-name='Default-notes' style:name='pr1' style:family='presentation'>
      <style:graphic-properties draw:fill-color='#ffffff' draw:auto-grow-height='true' fo:min-height='12.573cm' />
    </style:style>
    <style:style style:parent-style-name='Default-notes' style:name='pr2' style:family='presentation'>
      <style:graphic-properties fo:min-width='16.771cm' draw:fill-color='#ffffff' draw:auto-grow-height='true' fo:min-height='12.573cm' />
    </style:style>
    <style:style style:name='P1' style:family='paragraph'>
      <style:paragraph-properties fo:text-align='start' />
      <style:text-properties style:font-size-complex='12pt' fo:font-family='&apos;Liberation Mono&apos;' style:font-family-generic='modern' style:font-size-asian='12pt' style:font-pitch='fixed' fo:font-size='12pt' />
    </style:style>
    <style:style style:name='P2' style:family='paragraph'>
      <style:text-properties fo:font-size='20pt' />
    </style:style>
    <style:style style:name='P3' style:family='paragraph'>
      <style:paragraph-properties fo:text-align='center' />
      <style:text-properties style:font-size-complex='20pt' style:font-size-asian='20pt' fo:font-size='20pt' />
    </style:style>
    <style:style style:name='P4' style:family='paragraph'>
      <style:paragraph-properties fo:text-align='center' />
    </style:style>
END
    for my $pt (5..20,25,30,40,50) {
	print CONTENT <<"END";
    <style:style style:name='M$pt' style:family='text'>
      <style:text-properties style:font-size-complex='${pt}pt' fo:font-family='&apos;Liberation Mono&apos;' style:font-family-generic='modern' style:font-size-asian='${pt}pt' style:font-pitch='fixed' fo:color='#000000' fo:font-size='${pt}pt' />
    </style:style>
    <style:style style:name='S$pt' style:family='text'>
      <style:text-properties style:font-size-complex='${pt}pt' fo:font-family='&apos;Liberation Sans&apos;' style:font-family-generic='modern' style:font-size-asian='${pt}pt' fo:color='#000000' fo:font-size='${pt}pt' />
    </style:style>
END
    }
    print CONTENT <<"END";
    <style:style style:name='Magenta' style:family='text'>
      <style:text-properties fo:color='#ff20c0' />
    </style:style>
    <style:style style:name='Lavender' style:family='text'>
      <style:text-properties fo:color='#b040f0' />
    </style:style>
    <style:style style:name='Red' style:family='text'>
      <style:text-properties fo:color='#ff0000' />
    </style:style>
    <style:style style:name='Orange' style:family='text'>
      <style:text-properties fo:color='#ff8000' />
    </style:style>
    <style:style style:name='Yellow' style:family='text'>
      <style:text-properties fo:color='#ffff00' />
    </style:style>
    <style:style style:name='Green' style:family='text'>
      <style:text-properties fo:color='#00c000' />
    </style:style>
    <style:style style:name='Cyan' style:family='text'>
      <style:text-properties fo:color='#00e0e0' />
    </style:style>
    <style:style style:name='Blue' style:family='text'>
      <style:text-properties fo:color='#0000e0' />
    </style:style>
    <style:style style:name='Purple' style:family='text'>
      <style:text-properties fo:color='#900090' />
    </style:style>
    <style:style style:name='Black' style:family='text'>
      <style:text-properties fo:color='#000000' />
    </style:style>
    <style:style style:name='Gray' style:family='text'>
      <style:text-properties fo:color='#808080' />
    </style:style>
    <style:style style:name='White' style:family='text'>
      <style:text-properties fo:color='#ffffff' />
    </style:style>
    <text:list-style style:name='L1'>
      <text:list-level-style-bullet text:level='1' text:bullet-char='●'>
        <style:list-level-properties />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='2' text:bullet-char='●'>
        <style:list-level-properties text:space-before='0.6cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='3' text:bullet-char='●'>
        <style:list-level-properties text:space-before='1.2cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='4' text:bullet-char='●'>
        <style:list-level-properties text:space-before='1.8cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='5' text:bullet-char='●'>
        <style:list-level-properties text:space-before='2.4cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='6' text:bullet-char='●'>
        <style:list-level-properties text:space-before='3cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='7' text:bullet-char='●'>
        <style:list-level-properties text:space-before='3.6cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='8' text:bullet-char='●'>
        <style:list-level-properties text:space-before='4.2cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='9' text:bullet-char='●'>
        <style:list-level-properties text:space-before='4.8cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
      <text:list-level-style-bullet text:level='10' text:bullet-char='●'>
        <style:list-level-properties text:space-before='5.4cm' text:min-label-width='0.6cm' />
        <style:text-properties fo:font-family='StarSymbol' style:use-window-font-color='true' fo:font-size='45%' />
      </text:list-level-style-bullet>
    </text:list-style>
  </office:automatic-styles>
  <office:body>
    <office:presentation>
END

my $page = 0;

use Cwd ();
for (@para) {
    $page++;
    my $verbatim = /^\s/;
    s/^    //mg if $verbatim;
    s/^`//mg;
    if (/\bi:(.*)/) {
      my $name = Cwd::abs_path("image/$1");
      (my $safename = $name) =~ s/\//__/g;
      system "cp $name $build/Pictures/$safename";
      print CONTENT <<"END";
      <draw:page draw:style-name='dp1' draw:name='page$page' draw:master-page-name='Default'>
        <office:forms form:automatic-focus='false' form:apply-design-mode='false' />
        <draw:frame svg:y='0cm' draw:style-name='gr4' svg:x='0cm' draw:layer='layout' svg:width='10.008cm' svg:height='8.001cm' draw:text-style-name='P4'>
          <draw:image xlink:href='Pictures/$safename' xlink:actuate='onLoad' xlink:show='embed' xlink:type='simple'>
            <text:p />
          </draw:image>
        </draw:frame>
      </draw:page>
END
    }
    elsif (/\bh:(.*)/) {
	my $link = $1;
	print CONTENT <<"END";
      <draw:page draw:style-name='dp3' draw:name='page$page' draw:master-page-name='Default'>
        <draw:frame svg:y='4cm' draw:style-name='takahashi' svg:x='5cm' draw:layer='layout' draw:text-style-name='P3'>
          <draw:text-box>
            <text:p text:style-name='P3'>
		<text:a xlink:href='http:$link'>http:$link</text:a>
            </text:p>
          </draw:text-box>
        </draw:frame>
      </draw:page>
END
    }
    elsif (/\bf:(.*)/) {
	my $file = $1;
	print CONTENT <<"END";
      <draw:page draw:style-name='dp3' draw:name='page$page' draw:master-page-name='Default'>
        <draw:frame svg:y='4cm' draw:style-name='takahashi' svg:x='5cm' draw:layer='layout' draw:text-style-name='P3'>
          <draw:text-box>
            <text:p text:style-name='P1'>
		<text:span text:style-name='M5'>
		<text:a xlink:href='file:$file'>$file</text:a>
		</text:span>
            </text:p>
          </draw:text-box>
        </draw:frame>
      </draw:page>
END
    }
    else {
	my @lines = split(/^/, $_);
	my $maxlen = 1;
	for my $line (@lines) {
	    $line =~ s/\t/        /g;
	    $line =~ s/( +)/"\xa0" x length($1)/ge;
	    my $temp = $line;
	    $temp =~ s/`?\b([a-zA-Z])<(.*?)>/$2/g;
	    $temp =~ s/`?\b([a-zA-Z])://g;
	    $temp =~ s/[\x{2000}-\x{ffff}]/XXX/g;
	    $maxlen = length($temp) if $maxlen < length($temp);
	}
	for my $line (@lines) {
	    $line =~ s/`?\b([a-zA-Z])<(.*?)>/<text:span text:style-name='$COLOR{$1}'>$2<\/text:span>/g;
	    $line =~ s/`?\b([a-zA-Z]):(.*)/<text:span text:style-name='$COLOR{$1}'>$2<\/text:span>/g;
	}
	if ($verbatim) {
	    print CONTENT <<"END";
      <draw:page draw:style-name='dp3' draw:name='page$page' draw:master-page-name='Default'>
        <draw:frame svg:y='2cm' draw:style-name='middling' svg:x='0.4cm' draw:layer='layout' svg:width='10.5cm' svg:height='3.639cm' draw:text-style-name='P1'>
          <draw:text-box>
END
	    my $t = 'M' . (
		$maxlen > 70 ? 5 :
		$maxlen > 50 ? 7 :
		$maxlen > 40 ? 9 :
		$maxlen > 30 ? 12 :
		$maxlen > 20 ? 14 :
		$maxlen > 10 ? 20 :
		30
	    );
	    for my $line (@lines) {
		print CONTENT <<"END";
            <text:p text:style-name='P1'>
              <text:span text:style-name='$t'>$line</text:span>
            </text:p>
END
	    }
	    print CONTENT <<"END";
          </draw:text-box>
        </draw:frame>
      </draw:page>
END
	}
	else {
	    print CONTENT <<"END";
      <draw:page draw:style-name='dp3' draw:name='page3' draw:master-page-name='Default'>
        <draw:frame svg:y='4cm' draw:style-name='takahashi' svg:x='5cm' draw:layer='layout' draw:text-style-name='P3'>
          <draw:text-box>
END
	    my $t = 'S' . (
		$maxlen > 70 ? 5 :
		$maxlen > 60 ? 7 :
		$maxlen > 50 ? 8 :
		$maxlen > 45 ? 9 :
		$maxlen > 40 ? 10 :
		$maxlen > 35 ? 11 :
		$maxlen > 28 ? 14 :
		$maxlen > 24 ? 17 :
		$maxlen > 20 ? 20 :
		$maxlen > 15 ? 25 :
		$maxlen > 10 ? 30 :
		$maxlen > 8 ? 35 : 40
	    );
	    for my $line (@lines) {
		print CONTENT <<"END";
            <text:p text:style-name='P3'>
              <text:span text:style-name='$t'>$line</text:span>
            </text:p>
END
	    }
	    print CONTENT <<"END";
          </draw:text-box>
        </draw:frame>
      </draw:page>
END
	}
    }
}

print CONTENT <<'END';
      <presentation:settings presentation:mouse-visible='false' />
    </office:presentation>
  </office:body>
</office:document-content>
END

system("chdir $build; zip -r -u ../stump.odp *");

}

1;
